hazelcast/src/main/java/com/hazelcast/internal/blackbox/impl/SensorInputAnnotationScanner.java

```diff
@@ -0,0 +1,169 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.blackbox.impl;
+
+import com.hazelcast.internal.blackbox.SensorInput;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.internal.blackbox.impl.FieldSensorInputFactories.registerFieldTypes;
+import static com.hazelcast.internal.blackbox.impl.MethodSensorInputFactories.registerMethodTypes;
+import static java.lang.String.format;
+
+class SensorInputAnnotationScanner {
+    private final Map<Class, FieldSensorInputFactory> fieldFactories = new HashMap<Class, FieldSensorInputFactory>();
+    private final Map<Class, MethodSensorInputFactory> methodFactories = new HashMap<Class, MethodSensorInputFactory>();
+    private final BlackboxImpl blackbox;
+
+    public SensorInputAnnotationScanner(BlackboxImpl blackbox) {
+        this.blackbox = blackbox;
+        registerFieldTypes(fieldFactories);
+        registerMethodTypes(methodFactories);
+    }
+
+    void scanFields(String parameterPrefix, final Object source) {
+        Field[] fields = source.getClass().getDeclaredFields();
+        for (final Field field : fields) {
+            SensorInput fieldSensorInput = field.getAnnotation(SensorInput.class);
+
+            if (fieldSensorInput == null) {
+                continue;
+            }
+
+            FieldSensorInputFactory inputFactory = getFieldSensorInputFactory(field.getType());
+            if (inputFactory == null) {
+                throw new IllegalArgumentException(format("SensorInputField '%s' is of an unhandled type", field));
+            }
+
+            field.setAccessible(true);
+
+            Object input = inputFactory.create(field, source);
+            String parameter = getFieldParameter(parameterPrefix, field, fieldSensorInput);
+            blackbox.register(parameter, input, source);
+        }
+    }
+
+    private String getFieldParameter(String parameterPrefix, Field field, SensorInput fieldSensorInput) {
+        String sensorName = field.getName();
+        if (!fieldSensorInput.name().equals("")) {
+            sensorName = fieldSensorInput.name();
+        }
+
+        return parameterPrefix + "." + sensorName;
+    }
+
+    private FieldSensorInputFactory getFieldSensorInputFactory(Class clazz) {
+        FieldSensorInputFactory factory = fieldFactories.get(clazz);
+        if (factory != null) {
+            return factory;
+        }
+
+        List<Class> flattenedClassHierarchy = new LinkedList<Class>();
```

Comment:
Sure. Since it isn't relevant for performance I picked one. I'm fine with the other.